README.md
# ElectricTrain
Mod for Factorio

This mod adds 3 levels of electric trains and 2 improved cargo/fluid wagons.
The trains are supplied by one or more power providers, which can be built freely anywhere.

|Train|Consumption|Speed|
|-|-|-|
| E-locomotive | 900 KW | 388 KM/H |
| E-lolomotive MK2 | 1200 KW | 432 KM/H |
| E-lolomotive MK3 | 1.5 MW | 518 KM/H |
| Space Train | 1.8 MW | 604 KM/H |

|Wagon|Capacity|
|-|-|
|Cargo wagon MK2|60 stacks|1500 Kg|
|Cargo wagon MK3|80 stacks|2000 Kg|
|Space Cargo wagon|80 stacks|1500 Kg|
|||
|Fluid wagon MK2|37,500|1500 Kg|
|Fluid wagon MK3|50,000|2000 Kg|
|Space Fluid wagon|50,000|1500 Kg|

|Provider|Power|
|-|-|
|Power provider|2 MW|


changelog.txt

---------------------------------------------------------------------------------------------------
Version: 1.2
Date: 20220-12-28
  Changes:
    - Added 4th Train and wagon tier 'The Space Train'.
    - Changed Acceleration Power and max speed of old Mks.
---------------------------------------------------------------------------------------------------
Version: 1.1.2
Date: 2020-12-19
  Bugfixes:
    - Fixed technology icon issue.
---------------------------------------------------------------------------------------------------
Version: 1.1.1
Date: 2020-12-18
  Translation:
    - Spanish translation. (Pull request by eloiaro5)
    - Catalan translation. (Pull request by eloiaro5)
---------------------------------------------------------------------------------------------------
Version: 1.1.0
Date: 2020-11-27
  Major Features:
    - Release 1.1.
---------------------------------------------------------------------------------------------------
Version: 1.0.0
Date: 2020-11-27
  Major Features:
    - Release 1.0.
---------------------------------------------------------------------------------------------------
Version: 0.18.1
Date: 2020-01-31
  Major Features:
    - Release 0.18.
---------------------------------------------------------------------------------------------------
Version: 0.17.29
Date: 2019-11-02
  Bugfixes:
    - Try to fix uncraftable Control Station and Current Collector. https://mods.factorio.com/mod/ElectricTrain/discussion/5da63511338655000dcee54f 
---------------------------------------------------------------------------------------------------
Version: 0.17.28
Date: 2019-08-31
  Bugfixes:
    - Change code for compatibility with Multiple Unit Train Control
---------------------------------------------------------------------------------------------------
Version: 0.17.27
Date: 2019-08-24
  Changes:
    - Removed the old power provider
---------------------------------------------------------------------------------------------------
Version: 0.17.23
Date: 2019-08-21
  Bugfixes:
    - Change code for compatibility with Multiple Unit Train Control
---------------------------------------------------------------------------------------------------
Version: 0.17.22
Date: 2019-07-02
  Bugfixes:
    - Fixed a tech issue.
---------------------------------------------------------------------------------------------------
Version: 0.17.21
Date: 2019-06-29
  Bugfixes:
    - Fixed a recipe issue.
---------------------------------------------------------------------------------------------------
Version: 0.17.20
Date: 2019-06-28
  Minor Features:
    - Implements a new Behavior.
    - Control Station is the new building that powers the trains.
    - For future updates, i will add more control stations to supply the trains (e.g. mk1 can supply 2 trains, mk2 4.....) and a GUI that makes that visible.
---------------------------------------------------------------------------------------------------
Version: 0.17.14
Date: 2019-05-31
  Bugfixes:
    - Fixed a table issue.
---------------------------------------------------------------------------------------------------
Version: 0.17.13
Date: 2019-03-31
  Changes:
    - Fixed a typo.
---------------------------------------------------------------------------------------------------
Version: 0.17.12
Date: 2019-03-30
  Changes:
    - Fixed a typo.
---------------------------------------------------------------------------------------------------
Version: 0.17.11
Date: 2019-03-30
  Changes:
    - Fixed a mistake.
---------------------------------------------------------------------------------------------------
Version: 0.17.10
Date: 2019-03-30
  Changes:
    - Change code for compatibility with Multiple Unit Train Control
  Translation:
    - French translation. (Pull request by kryshnar)
---------------------------------------------------------------------------------------------------
Version: 0.17.9
Date: 2019-03-29
  Changes:
    - Update for factorio 0.17.23. Renamed "electric-energy-accumulators-1" to "electric-energy-accumulators"
---------------------------------------------------------------------------------------------------
Version: 0.17.8
Date: 2019-03-28
  Bugfixes:
    - Attempt to fix the error. https://mods.factorio.com/mod/ElectricTrain/discussion/5c9cd4b1e95caf000b9fc903
---------------------------------------------------------------------------------------------------
Version: 0.17.7
Date: 2019-03-27
  Changes:
    - Fixed a mistake.
---------------------------------------------------------------------------------------------------
Version: 0.17.6
Date: 2019-03-27
  Changes:
    - Change code for compatibility with Automatic Train Fuel Stop
---------------------------------------------------------------------------------------------------
Version: 0.17.5
Date: 2019-03-21
  Changes:
    - Next attempt to fix the "item without a fuel value" error. https://mods.factorio.com/mod/ElectricTrain/discussion/5c7bd6de961aa5000cbe7693
    - Attempt to fix the performance issues.
---------------------------------------------------------------------------------------------------
Version: 0.17.4
Date: 2019-03-17
  Changes:
    - Next attempt to fix the "item without a fuel value" error. https://mods.factorio.com/mod/ElectricTrain/discussion/5c7bd6de961aa5000cbe7693
---------------------------------------------------------------------------------------------------
Version: 0.17.3
Date: 2019-03-10
  Changes:
    - Attempt to fix the "item without a fuel value" error. https://mods.factorio.com/mod/ElectricTrain/discussion/5c7bd6de961aa5000cbe7693
    - Fixed the version issue. You have to update it manually. I'm sorry :(
---------------------------------------------------------------------------------------------------
Version: 0.17.1
Date: 2019-03-01
  Major Features:
    - Release 0.17.
---------------------------------------------------------------------------------------------------
Version: 0.16.12
Date: 2018-10-26

control.lua
require("lib")

local anzLoc = 0
local anzControl = 0


function CallRemoteInterface()
	if remote.interfaces["FuelTrainStop"] then
		remote.call("FuelTrainStop", "exclude_from_fuel_schedule", "et-electric-locomotive-1")
		remote.call("FuelTrainStop", "exclude_from_fuel_schedule", "et-electric-locomotive-2")
		remote.call("FuelTrainStop", "exclude_from_fuel_schedule", "et-electric-locomotive-3")
		remote.call("FuelTrainStop", "exclude_from_fuel_schedule", "et-electric-locomotive-4")
	end
end

function Init()
	global = {}
	global.LocList = {}
	global.ControlList = {}	
end

function Load()
	CallRemoteInterface()

	anzLoc = table.count(global.LocList)
	anzControl = table.count(global.ControlList)	
end

function Reinitialize()
	global = global or {}
	global.LocList = global.LocList or {}
	global.ControlList = global.ControlList or {}
	
	Load()
end

function OnInit()
	Init()
	Load()
end
script.on_init(OnInit)

function OnLoad()
	Load()
end
script.on_load(OnLoad)

function OnConfigurationChanged(data)
	local modName = "ElectricTrain-Microhawk"
	if not IsModChanged(data,modName) then
		Load()
	else
		oldVersion = GetOldVersion(data,modName)
		newVersion = GetNewVersion(data,modName)

		if oldVersion == newVersion then
			Reinitialize()
		else
			Init()
		
			for _,surface in pairs(game.surfaces) do
				local trains = surface.find_entities_filtered{type="locomotive"}
				for _,train in pairs(trains) do
					if train.name:match("^et%-electric%-locomotive%-%d$") or train.name:match("^et%-electric%-locomotive%-%d%-mu$") then
						table.insert(global.LocList,{entity=train,provider=nil})
						train.burner.currently_burning = game.item_prototypes['et-electric-locomotive-fuel']
					end
				end	
			end
			
			anzLoc = table.count(global.LocList)
			
			for _,surface in pairs(game.surfaces) do
				local controls = surface.find_entities_filtered{type="electric-energy-interface"}
				for _,control in pairs(controls) do
					if control.name:match("^et%-control%-station%-%d$") then
						table.insert(global.ControlList,control)
					end
					if control.name:match("^et%-electric%-locomotive%-%d%-power$") or control.name:match("^et%-electric%-locomotive%-%d%-mu-power$") then
						control.destroy()
					end
				end	
			end
			
			anzControl = table.count(global.ControlList)
		end
	end
end
script.on_configuration_changed(OnConfigurationChanged)

function OnBuiltEntity(event)
	local entity = event.created_entity or event.entity
	if entity and entity.valid then
		if entity.name:match("^et%-control%-station%-%d$") then
			table.insert(global.ControlList,entity)
			anzControl = anzControl + 1
		elseif entity.type == "locomotive" then
			if entity.name:match("^et%-electric%-locomotive%-%d$") or entity.name:match("^et%-electric%-locomotive%-%d%-mu$") then 
			table.insert(global.LocList,{entity=entity,provider=nil})
			entity.burner.currently_burning = game.item_prototypes['et-electric-locomotive-fuel']
			anzLoc = anzLoc + 1
			end
		end
	end
end
script.on_event({defines.events.on_built_entity,defines.events.on_robot_built_entity,defines.events.script_raised_built},OnBuiltEntity)

function OnRemoveEntity(event)
	local entity = event.entity
	if entity and entity.valid then
		if entity.name:match("^et%-control%-station%-%d$") then		
			for i,control in pairs(global.ControlList) do
				if control == entity then
					for _,loc in pairs(global.LocList) do
						if loc.provider and loc.provider.valid then
							loc.provider.destroy()
						end
						loc.provider = nil
					end
					table.remove(global.ControlList,i)
					anzControl = anzControl - 1
					break
				end
			end
		elseif entity.type == "locomotive" then
			if entity.name:match("^et%-electric%-locomotive%-%d$") or entity.name:match("^et%-electric%-locomotive%-%d%-mu$") then 
				for i,loc in pairs(global.LocList) do
					if loc.entity == entity then
						if loc.provider and loc.provider.valid then
							loc.provider.destroy()
						end
						table.remove(global.LocList,i)
						anzLoc = anzLoc - 1
						break
					end
				end
			end
		end
	end
end
script.on_event({defines.events.on_pre_player_mined_item,defines.events.on_robot_pre_mined,defines.events.on_entity_died,defines.events.script_raised_destroy},OnRemoveEntity)

function CreateProvider(loc)
	local control = global.ControlList[1]
	local pos = control.position
	local surface = control.surface
	local force = control.force
	local entity = surface.create_entity{name=loc.entity.name.."-power",position=pos,force=force}
	loc.provider = entity
end

function RemoveLoc(i)
	if global.LocList[i] then
		if global.LocList[i].entity and global.LocList[i].entity.valid then
			global.LocList[i].entity.destroy()
		end
		if global.LocList[i].provider and global.LocList[i].provider.valid then
			global.LocList[i].provider.destroy()
		end
	end
	table.remove(global.LocList,i)
end

function OnTick()
	if anzLoc > 0 and anzControl > 0 then
		for i,loc in pairs(global.LocList) do
			if loc and loc.entity and loc.entity.valid then
				if not (loc.provider and loc.provider.valid) then
					CreateProvider(loc)
				else	
					needPower = loc.entity.burner.currently_burning.fuel_value - loc.entity.burner.remaining_burning_fuel
					
					restPower = loc.provider.energy - needPower
					if restPower > 0 then
						loc.entity.burner.remaining_burning_fuel = loc.entity.burner.currently_burning.fuel_value
						loc.provider.energy = loc.provider.energy - needPower
					else
						loc.entity.burner.remaining_burning_fuel = loc.entity.burner.remaining_burning_fuel + loc.provider.energy
						loc.provider.energy = 0
					end
				end
			else
				RemoveLoc(i)
			end
		end
	end
end
--script.on_event(defines.events.on_tick,OnTick)
script.on_nth_tick(2,OnTick)

prototypes/item.lua
local electric_locomotive_1 = table.deepcopy(data.raw['item-with-entity-data']['locomotive'])
electric_locomotive_1.name = "et-electric-locomotive-1"
electric_locomotive_1.icon = "__SeElectricTrain__/graphics/loc1.png"
electric_locomotive_1.subgroup = "electric-transport-loc"
electric_locomotive_1.order = "b"
electric_locomotive_1.place_result = "et-electric-locomotive-1"
electric_locomotive_1.icon_size = 32
electric_locomotive_1.icon_mipmaps = nil

local electric_locomotive_2 = table.deepcopy(data.raw['item-with-entity-data']['locomotive'])
electric_locomotive_2.name = "et-electric-locomotive-2"
electric_locomotive_2.icon = "__SeElectricTrain__/graphics/loc2.png"
electric_locomotive_2.subgroup = "electric-transport-loc"
electric_locomotive_2.order = "c"
electric_locomotive_2.place_result = "et-electric-locomotive-2"
electric_locomotive_2.icon_size = 32
electric_locomotive_2.icon_mipmaps = nil

local electric_locomotive_3 = table.deepcopy(data.raw['item-with-entity-data']['locomotive'])
electric_locomotive_3.name = "et-electric-locomotive-3"
electric_locomotive_3.icon = "__SeElectricTrain__/graphics/loc3.png"
electric_locomotive_3.subgroup = "electric-transport-loc"
electric_locomotive_3.order = "d"
electric_locomotive_3.place_result = "et-electric-locomotive-3"
electric_locomotive_3.icon_size = 32
electric_locomotive_3.icon_mipmaps = nil

local electric_locomotive_4 = {
  type = "item-with-entity-data",
  name = "et-electric-locomotive-4",
  icon = "__SeElectricTrain__/graphics/icons/space-locomotive.png",
  icon_size = 64,
  subgroup = "electric-transport-loc",
  order = "e",
  place_result = "et-electric-locomotive-4",
  stack_size = 5,
  icon_mipmaps = nil
}

data:extend({electric_locomotive_1,electric_locomotive_2,electric_locomotive_3,electric_locomotive_4})


local cargo_wagon_2 = table.deepcopy(data.raw['item-with-entity-data']['cargo-wagon'])
cargo_wagon_2.name = "et-cargo-wagon-2"
cargo_wagon_2.icon = "__SeElectricTrain__/graphics/cargo2.png"
cargo_wagon_2.subgroup = "electric-transport-cargo"
cargo_wagon_2.order = "b"
cargo_wagon_2.place_result = "et-cargo-wagon-2"
cargo_wagon_2.icon_size = 32
cargo_wagon_2.icon_mipmaps = nil

local cargo_wagon_3 = table.deepcopy(data.raw['item-with-entity-data']['cargo-wagon'])
cargo_wagon_3.name = "et-cargo-wagon-3"
cargo_wagon_3.icon = "__SeElectricTrain__/graphics/cargo3.png"
cargo_wagon_3.subgroup = "electric-transport-cargo"
cargo_wagon_3.order = "c"
cargo_wagon_3.place_result = "et-cargo-wagon-3"
cargo_wagon_3.icon_size = 32
cargo_wagon_3.icon_mipmaps = nil

local cargo_wagon_4 = {
  type = "item-with-entity-data",
  name = "et-cargo-wagon-4",
  icon = "__SeElectricTrain__/graphics/icons/space-cargo-wagon.png",
  icon_size = 64,
  icon_mipmaps = nil,
  subgroup = "electric-transport-cargo",
  order = "d",
  place_result = "et-cargo-wagon-4",
  stack_size = 5
}

data:extend({cargo_wagon_2,cargo_wagon_3,cargo_wagon_4})	


local fluid_wagon_2 = table.deepcopy(data.raw['item-with-entity-data']['fluid-wagon'])
fluid_wagon_2.name = "et-fluid-wagon-2"
fluid_wagon_2.icon = "__SeElectricTrain__/graphics/fluid2.png"
fluid_wagon_2.subgroup = "electric-transport-fluid"
fluid_wagon_2.order = "b"
fluid_wagon_2.place_result = "et-fluid-wagon-2"
fluid_wagon_2.icon_size = 32
fluid_wagon_2.icon_mipmaps = nil

local fluid_wagon_3 = table.deepcopy(data.raw['item-with-entity-data']['fluid-wagon'])
fluid_wagon_3.name = "et-fluid-wagon-3"
fluid_wagon_3.icon = "__SeElectricTrain__/graphics/fluid3.png"
fluid_wagon_3.subgroup = "electric-transport-fluid"
fluid_wagon_3.order = "c"
fluid_wagon_3.place_result = "et-fluid-wagon-3"
fluid_wagon_3.icon_size = 32
fluid_wagon_3.icon_mipmaps = nil

local fluid_wagon_4 = {
  type = "item-with-entity-data",
  name = "et-fluid-wagon-4",
  icon = "__SeElectricTrain__/graphics/icons/space-fluid-wagon.png",
  icon_size = 64,
  icon_mipmaps = nil,
  subgroup = "electric-transport-fluid",
  order = "d",
  place_result = "et-fluid-wagon-4",
  stack_size = 5
};

data:extend({fluid_wagon_2,fluid_wagon_3,fluid_wagon_4})


data:extend(
{
	{
		type = "item",
		name = "et-electric-locomotive-fuel",
		icon = "__base__/graphics/icons/wood.png",
		icon_size = 32,
		flags = { "hidden" },
		fuel_value = "200kJ",
		fuel_category = "chemical",
		stack_size = 1
	}
})


local control_station_1 = table.deepcopy(data.raw['item']['small-lamp'])
control_station_1.name = "et-control-station-1"
control_station_1.icon = "__SeElectricTrain__/graphics/relais-icon-1.png"
control_station_1.subgroup = "electric-transport-basic"
control_station_1.order = "b"
control_station_1.place_result = "et-control-station-1"
control_station_1.icon_size = 32
control_station_1.icon_mipmaps = nil

data:extend({control_station_1})


data:extend(
{
	{
		type = "item",
		name = "et-current-collector",
		icon = "__SeElectricTrain__/graphics/current-collector-icon.png",
		icon_size = 32,
		subgroup = "electric-transport-basic",
		order = "a",
		stack_size = 200
	}
})


data:extend(
{
	{
		type = "item",
		name = "et-rocket-part",
		icon = "__base__/graphics/icons/rocket-part.png",
		icon_size = 32,
		flags = {"hidden"},
		subgroup = "intermediate-product",
		order = "q[et-rocket-part]",
		stack_size = 100
	}
})

data-final-fixes.lua
function UpdateMUControl(train,mu_power)	
	local power = format_number(train.max_power)	
	local energy = power * 1.1

	mu_power.icon = train.icon
	mu_power.localised_name = {"entity-name." .. train.name}
	mu_power.energy_source =
	{
		type = "electric",
		buffer_capacity = (energy * 2) .. "J",
		usage_priority = "secondary-input",
		input_flow_limit = energy .. "J" ,
		drain = power / 10 .. "J" ,
		render_no_network_icon = false,
		render_no_power_icon = false
	}
end

if mods['MultipleUnitTrainControl'] then
	UpdateMUControl(data.raw['locomotive']['et-electric-locomotive-1-mu'],data.raw['electric-energy-interface']['et-electric-locomotive-1-mu-power'])	
	UpdateMUControl(data.raw['locomotive']['et-electric-locomotive-2-mu'],data.raw['electric-energy-interface']['et-electric-locomotive-2-mu-power'])
	UpdateMUControl(data.raw['locomotive']['et-electric-locomotive-3-mu'],data.raw['electric-energy-interface']['et-electric-locomotive-3-mu-power'])
	UpdateMUControl(data.raw['locomotive']['et-electric-locomotive-4-mu'],data.raw['electric-energy-interface']['et-electric-locomotive-4-mu-power'])
end

prototypes/recipe.lua
data:extend(
{
	{
		type = "recipe",
		name = "et-electric-locomotive-1",
		enabled = false,
		ingredients =
		{
			{"locomotive", 1},
			{"et-current-collector", 2},
			{"battery", 10},
			{"electric-engine-unit", 10}
		},
		result = "et-electric-locomotive-1"
	},
	
	{
		type = "recipe",
		name = "et-control-station-1",
		enabled = false,
		ingredients =
		{
			{"electronic-circuit", 20},
			{"advanced-circuit", 20},
			{"steel-plate", 10},
			{"copper-cable", 10}
		},
		result = "et-control-station-1"
	},

	{
		type = "recipe",
		name = "et-current-collector",
		enabled = false,
		ingredients =
		{
			{"low-density-structure", 10},
			{"copper-cable", 5},
			{"iron-plate", 2}
		},
		result = "et-current-collector"
	}
})


function UnlockTrainRecipe()
	data:extend(
	{
		{	
			type = "recipe",
			name = "et-electric-locomotive-2",
			enabled = false,
			ingredients =
			{
				{"et-electric-locomotive-1", 1},
				{"low-density-structure", 10},
				{"electric-engine-unit", 10},
				{"advanced-circuit", 10}
			},
			result = "et-electric-locomotive-2"
		},
		
		{	
			type = "recipe",
			name = "et-electric-locomotive-3",
			enabled = false,
			ingredients =
			{
				{"et-electric-locomotive-2", 1},
				{"low-density-structure", 10},
				{"electric-engine-unit", 10},
				{"processing-unit", 10}	
			},
			result = "et-electric-locomotive-3"
		},
		
		{	
			type = "recipe",
			name = "et-electric-locomotive-4",
			energy_required = 60,
			enabled = false,
			ingredients =
			{
				{"et-electric-locomotive-3", 1},
				{"se-space-capsule", 1},
				{"se-cargo-rocket-section", 4},
			},
			result = "et-electric-locomotive-4"
		}
	})
end

function UnlockCargoRecipe()
	data:extend(
	{
		{	
			type = "recipe",
			name = "et-cargo-wagon-2",
			enabled = false,
			ingredients =
			{
				{"cargo-wagon", 1},
				{"iron-gear-wheel", 20},
				{"steel-plate", 20},
				{"low-density-structure", 10}	
			},
			result = "et-cargo-wagon-2"
		},
		
		{	
			type = "recipe",
			name = "et-cargo-wagon-3",
			enabled = false,
			ingredients =
			{
				{"et-cargo-wagon-2", 1},
				{"iron-gear-wheel", 20},
				{"steel-plate", 20},
				{"low-density-structure", 10}
			},
			result = "et-cargo-wagon-3"
		},
		
		{	
			type = "recipe",
			name = "et-cargo-wagon-4",
			enabled = false,
			ingredients =
			{
				{"et-cargo-wagon-3", 1},
				{"se-cargo-rocket-section", 2},
			},
			result = "et-cargo-wagon-4"
		},
	})
end

function UnlockFluidRecipe()
	data:extend(
	{
		{	
			type = "recipe",
			name = "et-fluid-wagon-2",
			enabled = false,
			ingredients =
			{
				{"fluid-wagon", 1},
				{"iron-gear-wheel", 10},
				{"steel-plate", 16},
				{"pipe", 8},
				{"storage-tank", 1},
				{"low-density-structure", 10}
			},
			result = "et-fluid-wagon-2"
		},
		
		{	
			type = "recipe",
			name = "et-fluid-wagon-3",
			enabled = false,
			ingredients =
			{
				{"et-fluid-wagon-2", 1},
				{"iron-gear-wheel", 10},
				{"steel-plate", 16},
				{"pipe", 8},
				{"storage-tank", 1},
				{"low-density-structure", 10}	
			},
			result = "et-fluid-wagon-3"
		},
		
		{	
			type = "recipe",
			name = "et-fluid-wagon-4",
			enabled = false,
			ingredients =
			{
				{"et-fluid-wagon-3", 1},
				{"se-cargo-rocket-section", 2},
			},
			result = "et-fluid-wagon-4"
		}
	})
end

prototypes/technology.lua
data:extend
({
	{
		type = "technology",
		name = "et-electric-railway",
		icon_size = 256,
		icon_mipmaps = 4,
		icon = "__SeElectricTrain__/graphics/electric-railway.png",
		effects =
		{
			{
				type = "unlock-recipe",
				recipe = "et-electric-locomotive-1"
			},
			{
				type = "unlock-recipe",
				recipe = "et-control-station-1"
			},
			{
				type = "unlock-recipe",
				recipe = "et-current-collector"
			}		
		},
		prerequisites = {"railway", "electric-engine", "battery", "low-density-structure"},
		unit =
		{
			count = 300,
			ingredients =
			{
				{"automation-science-pack", 1},
				{"logistic-science-pack", 1}
			},
			time = 30
		},
		order = "c-g-a-a"
	}
})

function UnlockTrainTechnology()
	data:extend
	({	
		{
			type = "technology",
			name = "et-electric-railway-2",
			icon_size = 256,
			icon_mipmaps = 4,
			icon = "__SeElectricTrain__/graphics/electric-railway.png",
			effects =
			{
				{
					type = "unlock-recipe",
					recipe = "et-electric-locomotive-2"
				}
			},
			prerequisites = {"et-electric-railway"},
			unit =
			{
				count = 300,
				ingredients =
				{
					{"automation-science-pack", 2},
					{"logistic-science-pack", 2},
					{"chemical-science-pack", 1},
				},
				time = 30
			},
			order = "c-g-a-a-a-b"
		},
	
		{
			type = "technology",
			name = "et-electric-railway-3",
			icon_size = 256,
			icon_mipmaps = 4,
			icon = "__SeElectricTrain__/graphics/electric-railway.png",
			effects =
			{
				{
					type = "unlock-recipe",
					recipe = "et-electric-locomotive-3"
				}
			},
			prerequisites = {"et-electric-railway-2", "utility-science-pack"},
			unit =
			{
				count = 300,
				ingredients =
				{
					{"automation-science-pack", 3},
					{"logistic-science-pack", 3},
					{"chemical-science-pack", 1},
					{"utility-science-pack", 1}
				},
				time = 30
			},
			order = "c-g-a-a-a-c"
		},
	})
end

function UnlockCargoTechnology()
	data:extend
	({	
		{
			type = "technology",
			name = "et-cargo-wagon-2",
			icon_size = 256,
			icon_mipmaps = 4,
			icon = "__base__/graphics/technology/railway.png",
			effects =
			{
				{
					type = "unlock-recipe",
					recipe = "et-cargo-wagon-2"
				}
			},
			prerequisites = {"railway", "low-density-structure"},
			unit =
			{
				count = 50,
				ingredients =
				{
					{"automation-science-pack", 3},
					{"logistic-science-pack", 2},
					{"chemical-science-pack", 1},
				},
				time = 30
			},
			order = "c-g-a-a-c"
		},
	
		{
			type = "technology",
			name = "et-cargo-wagon-3",
			icon_size = 256,
			icon_mipmaps = 4,
			icon = "__base__/graphics/technology/railway.png",
			effects =
			{
				{
					type = "unlock-recipe",
					recipe = "et-cargo-wagon-3"
				}
			},
			prerequisites = {"et-cargo-wagon-2", "utility-science-pack"},
			unit =
			{
				count = 100,
				ingredients =
				{
					{"automation-science-pack", 4},
					{"logistic-science-pack", 3},
					{"chemical-science-pack", 2},
					{"utility-science-pack", 1}
				},
				time = 30
			},
			order = "c-g-a-a-c-b"
		},
	})
end


function UnlockFluidTechnology()
	data:extend
	({	
		{
			type = "technology",
			name = "et-fluid-wagon-2",
			icon_size = 256,
			icon_mipmaps = 4,
			icon = "__base__/graphics/technology/fluid-wagon.png",
			effects =
			{
				{
					type = "unlock-recipe",
					recipe = "et-fluid-wagon-2"
				}
			},
			prerequisites = {"fluid-wagon", "low-density-structure"},
			unit =
			{
				count = 50,
				ingredients =
				{
					{"automation-science-pack", 3},
					{"logistic-science-pack", 2},
					{"chemical-science-pack", 1},
				},
				time = 30
			},
			order = "c-g-a-b-b"
		},
	
		{
			type = "technology",
			name = "et-fluid-wagon-3",
			icon_size = 256,
			icon_mipmaps = 4,
			icon = "__base__/graphics/technology/fluid-wagon.png",
			effects =
			{
				{
					type = "unlock-recipe",
					recipe = "et-fluid-wagon-3"
				}
			},
			prerequisites = {"et-fluid-wagon-2", "utility-science-pack"},
			unit =
			{
				count = 100,
				ingredients =
				{
					{"automation-science-pack", 4},
					{"logistic-science-pack", 3},
					{"chemical-science-pack", 2},
					{"utility-science-pack", 1}
				},
				time = 30
			},
			order = "c-g-a-b-c"
		},
	})
end

data-updates.lua
data.raw['item-with-entity-data']['locomotive'].subgroup = "electric-transport-loc"
data.raw['item-with-entity-data']['locomotive'].order = "a"
data.raw['item-with-entity-data']['cargo-wagon'].subgroup = "electric-transport-cargo"
data.raw['item-with-entity-data']['cargo-wagon'].order = "a"
data.raw['item-with-entity-data']['fluid-wagon'].subgroup = "electric-transport-fluid"
data.raw['item-with-entity-data']['fluid-wagon'].order = "a"


if settings.startup['mk-train'].value then
	UnlockTrainRecipe()
	UnlockTrainTechnology()
end

if mods["space-exploration"] then
  table.insert(data.raw["technology"]["se-space-rail"].effects, {
    type = "unlock-recipe",
    recipe = "et-electric-locomotive-4"
  })
  table.insert(data.raw["technology"]["se-space-rail"].effects, {
    type = "unlock-recipe",
    recipe = "et-fluid-wagon-4"
  })
  table.insert(data.raw["technology"]["se-space-rail"].effects, {
    type = "unlock-recipe",
    recipe = "et-cargo-wagon-4"
  })
end

if not (mods['boblogistics'] or mods['FactorioExtended-Trains']) then
	if settings.startup['mk-cargo'].value then
		UnlockCargoRecipe()
		UnlockCargoTechnology()
	end
	if settings.startup['mk-fluid'].value then
		UnlockFluidRecipe()
		UnlockFluidTechnology()
	end
	if settings.startup['mk-train'].value then
		data.raw['artillery-wagon']['artillery-wagon'].max_speed = 2.8
		if not settings.startup['mk-cargo'].value then
			data.raw['cargo-wagon']['cargo-wagon'].max_speed = 2.8
		end
		if not settings.startup['mk-fluid'].value then
			data.raw['fluid-wagon']['fluid-wagon'].max_speed = 2.8
		end
	end	
end


if mods['Vehicle Wagon'] and settings.startup['mk-train'].value then
	for _,wagon in pairs(data.raw['cargo-wagon']) do
		if wagon.name:match("vehicle%-wagon") then	
			wagon.max_speed = 2.8
		end
	end
end


if mods['FactorioExtended-Trains'] and settings.startup['mk-train'].value then
	data.raw['cargo-wagon']['cargo-wagon-2'].max_speed = 1.8
	data.raw['cargo-wagon']['cargo-wagon-3'].max_speed = 2.4
	data.raw['cargo-wagon']['cargo-wagon-4'].max_speed = 2.8
	data.raw['fluid-wagon']['fluid-wagon-2'].max_speed = 1.8
	data.raw['fluid-wagon']['fluid-wagon-3'].max_speed = 2.4
	data.raw['fluid-wagon']['fluid-wagon-4'].max_speed = 2.8
end


if mods['EvenMoreLight'] then
	for _,train in pairs(data.raw['locomotive']) do
		if train.name:match("^et%-electric%-locomotive%-%d$") then
			train.front_light =
				{
					{
						minimum_darkness = 0.3,
						intensity = 0.9,
						size = 60,
					},
					{
						minimum_darkness = 0.3,
						intensity = 0.9,
						size = 60,
					}
				}
			train.stand_by_light =
				{
					{
						minimum_darkness = 0.3,
						color = {b=1},
						shift = {-0.6, -3.5},
						size = 2,
						intensity = 0.5
					},
					{
						minimum_darkness = 0.3,
						color = {b=1},
						shift = {0.6, -3.5},
						size = 2,
						intensity = 0.5
					},
					{
						minimum_darkness = 0.3,
						intensity = 0.9,
						size = 60,
					},
					{
						minimum_darkness = 0.3,
						intensity = 0.9,
						size = 60,
					}
				}
		end				
	end
end



if mods['bobvehicleequipment'] then
	if settings.startup['bobmods-vehicleequipment-enablevehiclegrids'].value then
		data.raw['locomotive']['et-electric-locomotive-1'].equipment_grid = "bob-locomotive"
		if settings.startup['mk-train'] then
			data.raw['locomotive']['et-electric-locomotive-2'].equipment_grid = "bob-locomotive-2"
			data.raw['locomotive']['et-electric-locomotive-3'].equipment_grid = "bob-locomotive-3"
			data.raw['locomotive']['et-electric-locomotive-4'].equipment_grid = "bob-locomotive-3"
		end
		if settings.startup['mk-cargo'].value then
			data.raw['cargo-wagon']['et-cargo-wagon-2'].equipment_grid = "bob-cargo-wagon-2"
			data.raw['cargo-wagon']['et-cargo-wagon-3'].equipment_grid = "bob-cargo-wagon-3"
			data.raw['cargo-wagon']['et-cargo-wagon-4'].equipment_grid = "bob-cargo-wagon-3"
		end
	end
end


--if mods['reverse-factory'] then
--	for _,item in pairs(data.raw['item-with-entity-data']) do
--		if item.type == "locomotive" or item.type == "cargo-wagon" or item.type == "fluid-wagon" then
--			local recipe = data.raw['recipe'][item.name]
--			if recipe then
--				data:extend
--					({
--						{
--							type = "recipe",
--							name = "rf-" .. item.name,
--							category = "recycle",
--							hidden = true,
--							icon = item.icon,
--							subgroup = "rf-multiple-outputs",
--							energy_required = 30,
--							ingredients = {{item.name, 1}},
--							results = recipe.ingredients
--						}
--					})
--			end
--		end
--	end
--end